README.md
NewRelicData
============

This is a simple interface for accessing data through [NewRelic's RESTful API](https://newrelic.com/docs/docs/getting-started-with-the-new-relic-rest-api).

* [Gem on Rubygems.com](https://rubygems.org/gems/newrelic_data)
* [Documentation](http://jmervine.github.com/newrelic_data/doc/)
* [Coverage](http://jmervine.github.com/newrelic_data/coverage/)
* [History](https://github.com/jmervine/newrelic_data/blob/master/HISTORY.md)

### Installation

        gem install newrelic_data

Via Bundler

     sources :rubygems
     gem 'newrelic_data'

### Usage

     require 'newrelic_data'
     @newrelic = NewRelicData.new

     @newrelic.format = "json" # default
     
     @newrelic.configure({
         api_key:      "your_newrelic_api_key",
         account:      4321,     # NewRelic Account ID
         application:  654321,   # NewRelic Application ID 
         # Also supported:
         # application:  "app name",   # NewRelic Application name.
         query_params: {
             metrics:      [ "Apdex", "Apdex/home/index" ],
             field:        "score",
             begin:        @newrelic.today,
             end:          7.days_ago
         }
     })
     
     pp @newrelic.fetch

### Disclaimer

NewRelic (tm) is a trademark of [New Relic, Inc.](http://www.newrelic.com/). This was written without their consent and is in no way affiliated with them.

#### MIT License

Copyright (C) 2012 Joshua P. Mervine

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



HISTORY.md
NewRelicData History
====================

### 0.0.4

* Added n.hours_ago
* Added NewRelicData.date_formatter(...str...)

### 0.0.2

* Made work.
* Added support for "app_name" 
* Added functional tests.
  * `rspec ./spec/newrelic_data_functional.rb`

### 0.0.1

* Initial Release




spec/spec_helper.rb
require 'simplecov'
require 'pp'
SimpleCov.start do
    add_filter "/vendor/"
    add_filter "/coverage/"
    add_filter "/doc/"
end

require './lib/newrelic_data'

module Curl
  class Easy
    def self.perform str, &block
      data = DummyData.new
      yield(data)
      case str
      when /\.json/
        data.body_str = File.read('./spec/dummy_data.json')
      when /\.xml/
        data.body_str = File.read('./spec/dummy_data.xml')
      when /\.csv/
        data.body_str = File.read('./spec/dummy_data.csv')
      end
      return data
    end
  end
end

class DummyData
  attr_accessor :body_str, :headers
  def initialize 
    @headers = {}
  end
end


spec/functional_helper.rb
require 'pp'
require './lib/newrelic_data'


API_KEY = "SETME"
ACCT_ID = "SETME"
APP_ID  = "SETME"

if API_KEY == "SETME" or ACCT_ID == "SETME" or APP_ID == "SETME"
  abort "ABORT: Please ensure that you have set API_KEY, ACCT_ID and APP_ID in 'spec/functional_helper.rb' before running functional test."
end



spec/newrelic_data_spec.rb
require 'spec_helper'
describe NewRelicData do
  before(:all) do
    @n = NewRelicData.new
  end

  describe :new, "with params" do
    before(:all) do
      @set_test = {
        api_key:      "test_api_key",
        host:         "test.host.com",
        port:         8080,
        format:       "xml"
      }
    end
    
    it "should set params from passed options" do
      n = nil
      expect { n = NewRelicData.new @set_test }.to_not raise_error
      @set_test.each do |key,val|
        n.send(key).should eq @set_test[key]
      end
    end
  end

  # attributes with defaults
  [ :ssl, :host, :port, :format ].each do |attr|
    describe attr do
      it 'should have default value' do
        @n.send(attr).should_not be_nil
      end
      it 'should set from value' do
        @n.send("#{attr}=".to_sym, "foobar").should be
        @n.send(attr).should eq "foobar"
      end
    end
  end

  describe :query_string do
    it "return query param string for fetching data from NewRelic" do
      n = NewRelicData.new( account: 4321, application: 543210 )
      str = n.query_string
      str.should match /metrics\[\]=Apdex/
      str.should match /field=score/
      str.should match /app_id=543210/
    end
  end

  describe :fetch_path do
    it "return URI string for fetching data from NewRelic" do
      n = NewRelicData.new( account: 4321, application: 543210 )
      str = n.fetch_path
      str.should match /^https/
      str.should match /\.json/
      str.should match /accounts\/4321/
      str.should match /app_id=543210/
      str.should match /metrics\[\]=Apdex/
      str.should match /field=score/
    end
  end

  describe :fetch_raw do
    before(:all) do
      @nr = NewRelicData.new( account: 4321, application: 543210, api_key: "test_api_key" )
    end
    it "should fetch raw json" do
      @nr.format = "json"
      expect { JSON.parse(@nr.fetch_raw) }.to_not raise_error
    end
    it "should fetch raw xml" do
      @nr.format = "xml"
      expect { XmlSimple.xml_in(@nr.fetch_raw) }.to_not raise_error
    end
    it "should fetch raw csv" do
      @nr.format = "csv"
      expect { CSV.parse(@nr.fetch_raw) }.to_not raise_error
    end
  end

  describe :fetch do
    before(:all) do
      @nr = NewRelicData.new( account: 4321, application: 543210, api_key: "test_api_key" )
    end
    it "should fetch parsed json" do
      @nr.format = "json"
      @nr.fetch.should be_a Hash
      @nr.fetch.should have_key "foo"
    end
    it "should fetch parsed xml" do
      @nr.format = "xml"
      @nr.fetch.should be_a Hash
      @nr.fetch.should have_key "foo"
    end
    it "should fetch parsed csv" do
      @nr.format = "csv"
      @nr.fetch.should be_a Array
      @nr.fetch.first.first.should eq "foo"
    end
    it "should raise error when format is unknown" do
      @nr.format = "bad"
      expect { @nr.fetch }.to raise_error RuntimeError
    end
  end

  # attributes without defaults
  [ :api_key, :application, :account ].each do |attr|
    describe attr do
      it 'should raise error if not set' do
        expect { @n.send(attr) }.to raise_error
      end
      it 'should set from value' do
        @n.send("#{attr}=".to_sym, "foobar").should be
        @n.send(attr).should match /foobar/
      end
    end
  end

  describe :application do
    it "should set app_id in query_param" do
      @n.application = 54321
      @n.query_params.should have_key "app_id"
      @n.query_params.should_not have_key "app_name"
      @n.query_params["app_id"].should eq 54321
    end
    it "should set app_name in query_param" do
      @n.application = "foobar"
      @n.query_params.should have_key "app_name"
      @n.query_params.should_not have_key "app_id"
      @n.query_params["app_name"].should eq "foobar"
    end
  end

  describe :today do
    it "should return today" do
      DateTime.stub(:now).and_return(DateTime.parse("2012-06-07"))
      @n.today.should eq "2012-06-07T00:00:00Z"
    end
  end
  describe :yesterday do
    it "should return yesterday" do
      DateTime.stub(:now).and_return(DateTime.parse("2012-06-07"))
      @n.yesterday.should eq "2012-06-06T00:00:00Z"
    end
  end

  describe :date_formatter do
    it "should format date" do
      DateTime.stub(:now).and_return(DateTime.parse("2012-06-07 22:22:22"))
      @n.date_formatter(DateTime.now).should eq "2012-06-07T22:22:22Z"
    end
  end
  describe ".date_formatter" do
    it "should format date" do
      DateTime.stub(:now).and_return(DateTime.parse("2012-06-07 22:22:22"))
      NewRelicData.date_formatter(DateTime.now).should eq "2012-06-07T22:22:22Z"
    end
  end

end

describe Integer, "monkey patch" do
  describe :days_ago do
    (-5..5).each do |n|
      n = n*7
      it "it should return '#{n}' days ago" do
        DateTime.stub(:now).and_return(DateTime.parse("2012-06-07"))
        DateTime.parse(n.days_ago).should eq DateTime.parse("2012-06-07")-n
      end
    end
  end
  describe :hours_ago do
    (-5..5).each do |n|
      n = n*7
      it "it should return '#{n}' hours ago" do
        DateTime.stub(:now).and_return(DateTime.parse("2012-06-07 22:30"))
        Time.stub(:now).and_return(DateTime.parse("2012-06-07 22:30").to_time)
        DateTime.parse(n.hours_ago).should eq DateTime.parse("2012-06-07 22:00")-(n/24.0)
      end
    end
  end
end


lib/newrelic_data.rb
require 'date'
require 'curb'
require 'json'
require 'xmlsimple'
require 'csv'
require 'uri'

# A simple utility for getting data from NewRelic based
# loosly off of NewRelic's API.
class NewRelicData

  # Version for gem.
  VERSION = "0.0.4"

  # @return [String] NewRelic API key.
  attr_accessor :api_key
 
  # @return [Boolean] Use 'http' or 'https', defaults to true.
  attr_accessor :ssl

  # @return [String] NewRelic API host, defaults to 'rpm.newrelic.com'.
  attr_accessor :host
 
  # @return [Fixnum] NewRelic API port, defaults to '80'.
  attr_accessor :port
 
  # @return [Fixnum] NewRelic Account ID.
  attr_accessor :account
  
  # @return [Fixnum] NewRelic Application ID.
  attr_accessor :application
 
  # @return [String] Format to return NewRelic data; valid options:
  # * json
  # * xml
  # * csv
  attr_accessor :format
 
  # @return [Hash] Parameters to be passed to NewRelic API.
  attr_accessor :query_params

  # == Parameters:
  # See {#configure}.
  def initialize opts={}
    self.ssl = true
    configure(opts)
  end

  # == Parameters:
  # api_key::
  #   See {#api_key}
  # ssl::
  #   See {#ssl}
  # host:: 
  #   See {#host}
  # port::
  #   See {#port}
  # application::
  #   See {#application}
  # account::
  #   See {#account}
  # format::
  #   See {#format}
  # query_params::
  #   See {#query_params}
  #
  # == Example:
  #
  #     nr.configure( :query_params => { :metrics => [ "Apdex" ], :field => "score" },
  #                   :format => "xml",
  #                   :account => 4321, 
  #                   :application => 654321 )
  def configure opts
    opts.each do |key,val| 
      self.send("#{key}=".to_sym, val)
    end
  end

  # @return [Hash] or [Array] parsed data from NewRelic. 
  def fetch
    case self.format
    when "json"
      return JSON.parse(fetch_raw)
    when "xml"
      return XmlSimple.xml_in(fetch_raw)
    when "csv"
      return CSV.parse(fetch_raw)
    else
      raise "unsupport format '#{self.format}', try #fetch_raw"
    end
  end

  # @return [String] raw data from NewRelic.
  def fetch_raw
    Curl::Easy.perform(fetch_path) do |curl|
      curl.headers['x-api-key'] = self.api_key
    end.body_str
  end

  # @return [String] URI to be used to fetch content
  def fetch_path
    port = ":#{self.port}" unless self.port.nil? or self.port == 80
    "#{self.ssl ? 'https' : 'http' }://#{self.host}#{port}/api/v1/#{self.account}/metrics/data.#{self.format}?#{self.query_string}"
  end

  # @return [String] NewRelic API key.
  def api_key
    raise "api_key required" unless @api_key
    @api_key
  end

  # @return [String] NewRelic API host, defaults to 'api.newrelic.com'.
  def host
    @host||"api.newrelic.com"
  end

  # @return [Fixnum] NewRelic API port, defaults to '80'.
  def port
    @port||80
  end
 
  # @return [Fixnum] NewRelic Application ID.
  def application
    raise "application required" unless @application
    @application
  end

  def application=(a)
    @application = a
    if @application.kind_of? String and @application.to_i == 0
      self.query_params = self.query_params.merge({ "app_name" => @application })
      self.query_params.delete("app_id") if self.query_params.has_key?("app_id")
    else
      self.query_params = self.query_params.merge({ "app_id" => @application })
      self.query_params.delete("app_name") if self.query_params.has_key?("app_name")
    end
    @application
  end

  # @return [Fixnum] NewRelic Account ID.
  def account
    raise "account required" unless @account
    "accounts/#{@account}"
  end

  # @return [String] Format to return NewRelic data; valid options:
  # * json
  # * xml
  # * csv
  def format
    @format||"json"
  end

  # @return [Hash] Parameters to be passed to NewRelic API.
  #
  # == Parameters:
  # * "metrics" -- [Array]  (see NewRelic API)
  # * "field"   -- [String] (see NewRelic API)
  # * "begin"   -- [String] (format: YYYY-MM-DDTHH:MM:SSZ)
  # * "end"     -- [String] (format: YYYY-MM-DDTHH:MM:SSZ)
  #
  # == Note:
  # 'begin' and 'end' can be set using X.days_ago or helpers {#today} or {#yesterday}
  def query_params
    @query_params||{ "metrics" => [ "Apdex" ], "field" => "score", "begin" => 1.day_ago, "end" => self.today, "summary" => 1 }
  end

  def query_params=(q)
    @query_params = q
    @query_params["app_id"] = @application if @application and !@query_params.has_key?("app_id")
  end

  # @return [String] {#query_params} as a URI query string.
  def query_string
    q = []
    self.query_params.each do |key,val|
      if val.kind_of? Array
        val.each do |v|
          q.push "#{key}[]=#{v}"
        end
      else
        q.push "#{key}=#{val}"
      end
    end
    URI.escape(q.join("&"))
  end

  # @return [String] Today formated for 'begin' or 'end' {#query_params}.
  def today
    0.days_ago
  end

  # @return [String] Yesterday formated for 'begin' or 'end' {#query_params}.
  def yesterday
    1.day_ago
  end

  # @return [String] Formatted date / time
  def self.date_formatter(date_time)
    date_time.strftime("%Y-%m-%dT%H:%M:%SZ")
  end

  def date_formatter(date_time)
    self.class.date_formatter(date_time)
  end

end

# Monkey patch of the [Integer] object for quickly generating
# {NewRelicData#query_params} options 'begin' and 'end'.
class Integer
  # @return [String] formated for {NewRelicData#query_params} 'begin' and 'end'.
  # * N.days_ago at '00:00:00'
  def days_ago
    (DateTime.now-self).strftime("%Y-%m-%dT00:00:00Z")
  end
  alias :day_ago :days_ago
  def hours_ago
    (DateTime.now-self/24.0).strftime("%Y-%m-%dT%H:00:00Z")
  end
  alias :hour_ago :hours_ago
end

